### ISSUE TITLE
Advice on using LDpred2 - QC of GWAS, code, and difference with LDpred

### ISSUE THREAD

**Issue by QinqinHuang** [privefl/bigsnpr#241]
Hi Florian,

Thank you for making this tool and preparing the LD reference data and online tutorials, which are extremely helpful!

I have several questions that I want to discuss with you:

(1) Quality control of summary stats (Section 3.4 in your Bioinformatics paper).
Removing variants with sd_ss < (0.5 * sd_ldref) or sd_ss > (sd_ldref + 0.1) or sd_ss < 0.1 or sd_ldref < 0.05 is recommended. For **quantitative** traits (e.g. educational attainment measured by the number of years of schooling), I think we need to modify the code. The effect size and se have been standardised (page 10, Lee et al supplementary materials), based on equation (2) in the LDpred2 paper, I think sd_ss should be calculated as:
```sd_ss = with(info_snp, 1 / sqrt(n * beta_se^2))```
The code in your script (for binary traits only, I think): ```sd_ss = with(info_snp, 2 / sqrt(n_eff * beta_se^2))```
                Do you agree? I’m slightly confused by the bit about the maximum of sd_ss should be sqrt(0.5) in section 3.4…

PS the QC plot I got using the modified code looks fine:
![image](https://user-images.githubusercontent.com/10149527/125793228-8c350a5c-5e3e-42a9-900b-6a933cbc6cf6.png)

(2) heritability estimation in the “example-with-provided-ldref.R” script
I downloaded the R script along with the LD reference data that you shared on figshare. Thanks!
When calculating the heritability using LD score regression (line 74), the object “ld” was not defined anywhere in the script. Do you agree that I should add the following two lines before line 66 and line 68?
```
>65 if (chr == 1) {
      ld <- Matrix::colSums(corr_chr^2)
>66   corr <- as_SFBM(corr_chr, tmp)
>67  } else {
       ld <- c(ld, Matrix::colSums(corr_chr^2))
>68    corr$add_columns(corr_chr, nrow(corr))
>69  }
>70}
 ```
 
Then on line 74, it should be ```snp_ldsc(ld, ld_size = length(ld),```  rather than ```snp_ldsc(ld, ld_size = nrow(map_ldref),```
 
Do the modifications look correct to you?

(3) After I have done the QC on the summary data and calculated heritability (SCZ 0.363 & EA 0.115, which look fine), it freezes at the step ```beta_inf = snp_ldpred2_inf(corr, df_beta, h2 = h2_est)```. I couldn’t get any output. Is there anything that I can test? How long does this step usually take?
I am using R 3.6.1, and bigsnpr_1.8.1. I requested 15 cores and 10G RAM. Number of variants is round 1.0 million for both traits. It’s been running for more than 18 hours. The Rscript that I am using can be found [here](https://github.com/QinqinHuang/GEL_DDD_PRS).

(4) LDpred vs LDpred2
We plan to use the optimised hyper-parameters (the fraction of causal variants) of LDpred that was published previously, to avoid overfitting. Also, we don’t have the same phenotype in the validation set as in the original GWASs.
Would we get a similar PRS using the same p in LDpred2 with the sparsity disenabled? We will use the LDSC estimate for heritability. However, the window size parameter would be different between the two versions. LD reference panels might be slightly different, but I’m not sure how big the effects are if the reference panels are both European-ancestry. I can still use LDpred if you think there are too many uncertainties. Any comments or suggestions would be much appreciated!

**Comment by privefl** [comment]
1. Yes, replace `2` by `sd(y)` for continuous traits (which is 1 if standardized), and the QC plot looks fine. The max of the standard deviation that you can approximate by sqrt(2 * p * (1-p)) is reached at p = 0.5 and is therefore sqrt(0.5).

2. `ld` should be a precomputed column of `map_ldref`, so that you don't need to compute it again using a subset of the data.

3. LDpred2-inf should be quite fast (I would say max 10 seconds if the correlation matrix is already cached, or the time to read from the disk sbk file if not -> maybe a few minutes). You need to have enough memory such that the whole file fits in memory (should be ~30 GB) + some memory for the things you return (you should be fine with 60 GB). Note that I will announce soon a new format that should require half the memory and also be faster.

4. You should not use optimal parameters from LDpred1. LDpred1 will almost always return a high p as the optimal one, which is often not the case in LDpred2. If you want to avoid overfitting, given the QC plot which looks good, you could run LDpred2-auto instead.

**Comment by QinqinHuang** [comment]
Thanks for the quick reply! Very helpful.

1. Thanks for confirming. 
    Right, yes, thanks!

2. Ah I see! I'm sorry that I missed the ld column in ```map_ldref```. I'm wondering if we should use ```ld_size = nrow(df_beta)``` instead? there are fewer variants in ```df_beta``` (matched SNPs) than ```map_ldref``` (HM3 common SNPs)
```
ldsc <- with(df_beta, snp_ldsc(ld, ld_size = nrow(map_ldref),
                               chi2 = (beta / beta_se)^2,
                               sample_size = n_eff))
```

3. Thanks, I will test 60GB. 
    Great, Look forward to the new release.

4. Thank you! Sounds good. I will run LDpred2-auto too.

**Comment by privefl** [comment]
2. No I think you should provide the number of variants that were used to compute the LD scores.

3. The new release was announced on Twitter yesterday.

**Comment by QinqinHuang** [comment]
2. Thanks for the clarification! Yes, you are right, that should be the number of variants used to calculate the LD scores. Sorry, not very familiar with LDSC!

3. Great!

**Comment by QinqinHuang** [comment]
3: I have reinstalled the package. Calculating correlations is much quicker!
    ```snp_ldpred2_inf``` finished within 2 minutes.
    waiting for ```snp_ldpred2_grid``` and ```snp_ldpred2_auto``` 

Thank you very much!

**Comment by privefl** [comment]
You can close the issue when you think this is solved.

**Comment by QinqinHuang** [comment]
Hi Florian,

Thank you. I had a new issue when running LDpred-auto, something about "object '.doSnowGlobals' not found", but I think it's resolved now by adding my R library path to `$R_LIBS_USER` before bsub. `snp_ldpred2_auto` is running without complaining for a few hours now and I think it should be fine.

Thanks again for the amazing help!
